app/session.server.ts

```diff
@@ -43,9 +45,56 @@ export async function getUsername(request: Request): Promise<User['username'] |
   return username;
 }
 
+export async function getEffectiveUsername(
+  request: Request
+): Promise<User['username'] | undefined> {
+  const cookie = request.headers.get('Cookie');
+  const effectiveUsername = await effectiveUsernameCookie.parse(cookie);
+  return effectiveUsername as User['username'];
+}
+
+export async function getEffectiveUser(request: Request) {
+  // get the effective username in the cookie at the time
+  const [username, effectiveUsername] = await Promise.all([
+    getUsername(request),
+    getEffectiveUsername(request),
+  ]);
+
+  if (!username) {
+    return null;
+  }
+
+  // if the user doesn't have an effective username,
+  // we're using their normal username (not impersonating)
+  let user;
+  if (!effectiveUsername) {
+    user = await getUserByUsername(username);
+  } else {
+    user = await getUserByUsername(effectiveUsername);
+  }
+
+  if (user) {
+    return user;
+  }
+
+  throw await logout(request);
+}
+
+export async function setEffectiveUsername(
+  username: string,
+  name: string | undefined
```

Comment:
nit: you can use `name?: string` to mean "this is a `string` or undefined".

Reply:
Should `name` really be called `effectiveUsername`?  I think using `name` is ambiguous.